### PR TITLE
Add the ability to view Invidious playlists

### PR DESCRIPTION
**Pull request by ChunkyProgrammer**
# Add the ability to view Invidious playlists

## Pull Request Type
- [x] Feature Implementation

## Description
This PR allows us to view Invidious playlists (playlists that were created on Invidious)

## Testing 
- enter this playlist in the search bar: https://vid.puffyan.us/playlist?list=IVPLl_MgDeoZc0LTKRPutqlmivaY8btU171
- see videos in playlist (even if using local api or different invidious instance is set)
- share playlist (see only IV options) 
image
- play a video from playlist
- wtach-video-playlist fills with videos from the IV playlist

## Desktop
- **OS:** Windows
- **OS Version:** 10
- **FreeTube version:** 0.19.1

### CODE REVIEW

**Comment by absidue** [comment]
When the local API is selected you can only support these playlists if backend fallback is enabled, please add the relevant checks. Feel free to take inspiration from how we hide the popular button in the side bar.

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
> NOTE: THIS PR SHOULD BE MERGED AFTER: #4234 AND #4242

Could u maybe resolve conflicts so i can do a proper review :)

**Comment by ChunkyProgrammer** [comment]
Will wait for https://github.com/FreeTubeApp/FreeTube/pull/4597 to be merged before adding rest of changes (basically the playlist view loads but the playlist on the watch page still has some issues)

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
Uhh...can someone explain how this happened. Fetched the playlist with IV instance that doesnt exist.

https://github.com/FreeTubeApp/FreeTube/assets/73130443/a829b70c-87e5-47d9-b8db-7f7fb566a69c

**Comment by ChunkyProgrammer** [comment]
> Uhh...can someone explain how this happened. Fetched the playlist with IV instance that doesnt exist.
> 
>  VirtualBoxVM_sDUm79wFxG.mp4 
> Edit: Oh wait this doesnt use the API to fetch the playlist so it doesnt matter, right?

Yeah, the invidious playlists only exist on the imstance they are created so it makes a request to that instance instead of the configured one. The images still get proxied through your configured instance which is why the images dont load

**Review by efb4f5ff-1298-471a-8973-3d47447115dc — COMMENTED** [review]
Probably forget to localize this

Screenshot (325)

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
I only see 100 vids in the playlist instead of the 144, is this intended behavior?

Sorry for the comment spam

**Comment by ChunkyProgrammer** [comment]
> I only see 100 vids in the playlist instead of the 144, is this intended behavior?

This is because we dont support pagination for invidious playlists yet

**Comment by absidue** [comment]
Reminder for why we don't support pagination with Invidious, because they let you request more pages than actually exist and just return stuff from previous pages, so you don't know when you've reached the end.

**Comment by ChunkyProgrammer** [comment]
> Reminder for why we don't support pagination with Invidious, because they let you request more pages than actually exist and just return stuff from previous pages, so you don't know when you've reached the end.

This should be _mostly_ fixed since https://github.com/iv-org/invidious/pull/3807 (there might be one or duplicate videos). So we can probably add pagination support for invidious playlists at some point

**Review by efb4f5ff-1298-471a-8973-3d47447115dc — APPROVED** [review]
LGTM!

**Comment by PikachuEXE** [comment]
I can't even view the playlist? (local API)
image

It works in IV API but is that expected to not work in local API?

**Comment by ChunkyProgrammer** [comment]
> I can't even view the playlist? (local API) ![image](https://private-user-images.gi
> It works in IV API but is that expected to not work in local API?


Do you have backend fallback enabled?
https://github.com/FreeTubeApp/FreeTube/pull/4299#issuecomment-1801211183

**Comment by PikachuEXE** [comment]
OK now enabled but playlist does not exist
`The currently set default instance is https://invidious.io.lol`
image

**Comment by PikachuEXE** [comment]
Saw some missing videos, expected?
image
image

**Comment by ChunkyProgrammer** [comment]
> Saw some missing videos, expected? 

Not expected but I don't think it's caused by this PR

File: src/renderer/components/ft-share-button/ft-share-button.js [hunk 1]
```diff
@@ -38,6 +38,10 @@ export default defineComponent({
     dropdownPositionY: {
       type: String,
       default: 'bottom'
+    },
+    invidiousInstance: {
+      type: String,
+      default: null
     }
```
Comment (PikachuEXE) [inline src/renderer/components/ft-share-button/ft-share-button.js:45]:
```suggestion
    },
```

File: src/renderer/components/ft-share-button/ft-share-button.js [hunk 2]
```diff
@@ -141,6 +155,16 @@ export default defineComponent({
...
+
+    shareTitle: function() {
+      if (this.isChannel) {
+        return this.$t('Share.Share Channel')
+      }
+      if (this.isPlaylist || this.isIVPlaylist) {
+        return this.$t('Share.Share Playlist')
+      }
+      return this.$t('Share.Share Video')
+    }
```
Comment (PikachuEXE) [inline src/renderer/components/ft-share-button/ft-share-button.js:167]:
```suggestion
    },
```

File: src/renderer/components/watch-video-playlist/watch-video-playlist.js [hunk 3]
```diff
@@ -39,6 +39,14 @@ export default defineComponent({
...
       required: true,
     },
+    isInvidiousPlaylist: {
+      type: Boolean,
+      default: false,
+    },
+    origin: {
+      type: String,
+      default: null,
+    }
```
Comment (PikachuEXE) [inline src/renderer/components/watch-video-playlist/watch-video-playlist.js:49]:
```suggestion
    },
```

File: src/renderer/helpers/api/invidious.js [hunk 4]
```diff
@@ -78,10 +79,11 @@ export async function invidiousGetChannelInfo(channelId) {
   })
 }
 
-export async function invidiousGetPlaylistInfo(playlistId) {
+export async function invidiousGetPlaylistInfo(playlistId, origin = '') {
   return await invidiousAPICall({
     resource: 'playlists',
     id: playlistId,
+    instance: origin
```
Comment (PikachuEXE) [inline src/renderer/helpers/api/invidious.js:86]:
```suggestion
    instance: origin,
```

File: src/renderer/views/Playlist/Playlist.js [hunk 5]
```diff
@@ -53,14 +54,17 @@ export default defineComponent({
...
       playlistItems: [],
       userPlaylistVisibleLimit: 100,
       continuationData: null,
       isLoadingMore: false,
       getPlaylistInfoDebounce: function() {},
       playlistInEditMode: false,
-
       promptOpen: false,
+      query: {}
+
```
Comment (PikachuEXE) [inline src/renderer/views/Playlist/Playlist.js:67]:
```suggestion
      query: {},
```

**Review by PikachuEXE — COMMENTED** [review]
Trailing commas related
I will commit these changes

**Review by PikachuEXE — APPROVED** [review]

**Comment by PikachuEXE** [comment]
@efb4f5ff-1298-471a-8973-3d47447115dc Sorry files updated for formatting you need to re-approve 😬

**Review by efb4f5ff-1298-471a-8973-3d47447115dc — DISMISSED** [review]

**Comment by PikachuEXE** [comment]
You got some strange style in playlist view for all playlist types 
image

**Comment by ChunkyProgrammer** [comment]
@efb4f5ff-1298-471a-8973-3d47447115dc the playlist only exists on https://vid.puffyan.us . I added the view playlist button if the current invidious instance is different than the one the playlist is on so it won't auto fetch from other instances

**Review by efb4f5ff-1298-471a-8973-3d47447115dc — COMMENTED** [review]
After you click on view playlist you momentarily will see an empty playlist before it loads the actual playlist

https://github.com/FreeTubeApp/FreeTube/assets/73130443/7db22aa4-51e1-4130-bb0f-1aaa91a781dd

Im getting this error when going to the next/previous video in the playlist or if i use the navigation arrows in the top left corner to go back/forward

https://github.com/FreeTubeApp/FreeTube/assets/73130443/2fd43d59-a99f-429f-8c86-6bcbb285987c

it should load the playlist instead of showing this message

Capture

My fallback is the instance it is stored on so it should just let me through. Just like it did when viewing the playlist

https://github.com/FreeTubeApp/FreeTube/assets/73130443/da1a0589-ded5-4c71-ac48-7bfc345eb447

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
All issues are solved expect for

> Im getting this error when going to the next/previous video in the playlist or if i use the navigation arrows in the top left corner to go back/forward

https://github.com/FreeTubeApp/FreeTube/assets/73130443/2fd43d59-a99f-429f-8c86-6bcbb285987c

**Comment by ChunkyProgrammer** [comment]
@efb4f5ff-1298-471a-8973-3d47447115dc I'm getting that error in the development branch so it seems to be unrelated to this PR

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
> Edit: Can we get the same spacing on both sides of the big grey icon?

Only thing left to do

**Comment by ChunkyProgrammer** [comment]
> > Edit: Can we get the same spacing on both sides of the big grey icon?
> 
> Only thing left to do

Done @efb4f5ff-1298-471a-8973-3d47447115dc

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
Getting an error :(

https://github.com/FreeTubeApp/FreeTube/assets/73130443/7116dab0-625c-4204-8680-2ae2652f4b80

**Comment by ChunkyProgrammer** [comment]
> Getting an error :(
> VirtualBoxVM_ToocypTQTi.mp4

I think that's a bug in this PR: 
https://github.com/FreeTubeApp/FreeTube/pull/4537

I can reproduce that error on `development` by directly navigating to this playlist:
 https://youtube.com/playlist?list=PLJ-qODNIUEEtPdKZNLfbx7JOuRA_JjUxI

**Comment by ChunkyProgrammer** [comment]
Did you directly paste the playlist link in the navigation bar?
I get this on the latest development branch
image

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
I pasted it directly yes, im using `v0.19.1-nightly-4103 Beta`

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
Hmm weird dont see it on the latest nightly but i do see it on the dev branch

**Review by efb4f5ff-1298-471a-8973-3d47447115dc — DISMISSED** [review]

**Comment by PikachuEXE** [comment]
To fix the error simply update `searchQueryTextRequested()` in `src/renderer/views/Playlist/Playlist.js` to
```js
    searchQueryTextRequested() {
      return this.$route.query.searchQueryText ?? ''
    },
```

Coz `playlist-info` expect string but this query can return `undefined`

**Review by efb4f5ff-1298-471a-8973-3d47447115dc — APPROVED** [review]

**Review by PikachuEXE — DISMISSED** [review]

File: src/renderer/views/Watch/Watch.js [hunk 6]
```diff
@@ -1186,11 +1188,15 @@ export default defineComponent({
...
       this.playlistItemId = this.$route.query.playlistItemId
+      this.isInvidiousPlaylist = this.$route.query.playlistType === 'invidious'
+      this.origin = this.$route.query.origin
 
       if (this.playlistId == null || this.playlistId.length === 0) {
         this.playlistType = ''
         this.playlistItemId = null
         this.watchingPlaylist = false
+        this.playlistId = this.$route.query.playlistId
+        this.playlistItemId = this.$route.query.playlistItemId
```
Comment (absidue) [inline src/renderer/views/Watch/Watch.js:1199]: These two lines look identical to the lines, only a few lines further up. Are these needed?

**Comment by absidue** [comment]
Looks like the share menu on the watch page doesn't work correctly. For the YouTube links it includes the playlist ID, even though it it won't work as it's an Invidious playlist and for the Invidious ones it uses the current instance instead of the playlist's instance.

**Comment by ChunkyProgrammer** [comment]
> Looks like the share menu on the watch page doesn't work correctly. For the YouTube links it includes the playlist ID, even though it it won't work as it's an Invidious playlist and for the Invidious ones it uses the current instance instead of the playlist's instance.

I guess we should just remove the playlist id in this case?

**Comment by ChunkyProgrammer** [comment]
I will be setting this to WIP as I'd like to revisit this once the piped PR is merged (I'd also like to add support for piped playlists)

**Comment by ChunkyProgrammer** [comment]
Closing for now, I'll reopen once the Piped stuff is merged

**Pull request closed without merging**